pkg/telemetry/storage/db.go

```diff
@@ -0,0 +1,108 @@
+package storage
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/XSAM/otelsql"
+	"github.com/golang-migrate/migrate/v4"
+	_ "github.com/golang-migrate/migrate/v4/database/sqlite3" // migration for sqlite3
+	bindata "github.com/golang-migrate/migrate/v4/source/go_bindata"
+
+	"github.com/rs/zerolog"
+	logger "github.com/rs/zerolog/log"
+	"github.com/textileio/go-tableland/pkg/telemetry"
+	"github.com/textileio/go-tableland/pkg/telemetry/storage/migrations"
+	"go.opentelemetry.io/otel/attribute"
+)
+
+// TelemetryDatabase implements the MetricStore interface and provides storage for a metric.
+type TelemetryDatabase struct {
+	log   zerolog.Logger
+	sqlDB *sql.DB
+}
+
+// New returns a new TelemetryDatabase backed by database/sql.
+func New(dbURI string) (*TelemetryDatabase, error) {
+	sqlDB, err := otelsql.Open("sqlite3", dbURI, otelsql.WithAttributes(
+		attribute.String("name", "telemetrydb"),
+	))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to db: %s", err)
+	}
+	sqlDB.SetMaxIdleConns(0)
+	if err := otelsql.RegisterDBStatsMetrics(sqlDB, otelsql.WithAttributes(
+		attribute.String("name", "telemetrydb"),
+	)); err != nil {
+		return nil, fmt.Errorf("registering dbstats: %s", err)
+	}
+
+	log := logger.With().
+		Str("component", "telemetrydb").
+		Logger()
+
+	db := &TelemetryDatabase{
+		log:   log,
+		sqlDB: sqlDB,
+	}
+
+	as := bindata.Resource(migrations.AssetNames(), migrations.Asset)
+	if err := db.executeMigration(dbURI, as); err != nil {
+		return nil, fmt.Errorf("initializing db connection: %s", err)
+	}
+
+	return db, nil
+}
+
+// StoreMetric persists a metric.
+func (db *TelemetryDatabase) StoreMetric(ctx context.Context, metric telemetry.Metric) error {
+	payloadJSON, err := metric.Serialize()
+	if err != nil {
+		return fmt.Errorf("marshal json: %s", err)
+	}
+
+	_, err = db.sqlDB.ExecContext(ctx,
+		`INSERT INTO system_metrics ("timestamp", "type", "payload", "published") VALUES (?1, ?2, ?3, ?4)`,
+		metric.Timestamp.UnixMilli(), metric.Type, payloadJSON, 0,
```

Comment:
I thought `seconds` was not a good enough granularity for the timestamp of a metric. Is that thinking correct?

Reply:
Yeah, forget what I said, that's fine. The thing is that we can't even trust this timestamp since comes from a place we don't know, but at least in relative terms milliseconds can give us relative order which is fine. Not sure we would need to distinguish ordering at the millisecond level for a metric but who knows.
Sounds fine.